synapse/handlers/sso.py

```diff
@@ -702,8 +710,100 @@ async def _register_mapped_user(
+
         return registered_user_id
 
+    async def set_avatar(self, user_id: str, picture_https_url: str) -> None:
+        """Set avatar of the user.
+
+        This downloads the image file from the url provided, store that in
+        the media repository and then sets the avatar on user's profile.
+
+        It can detect if the same image is being saved again and bails early by storing
+        hash of the file in upload_name of the avatar image.
+
+        Currently, it only supports server configurations which runs media repository
+        within the same process.
+
+        It silently fails and log a warning by raising exception & catching it internally
+        if it is unable to fetch the image itself (non 200 status code) or
+        if the image supplied is bigger than max allowed size or
+        if the image type is not one of the allowed image types.
+
+        Args:
+            user_id: matrix user ID in the form @localpart:domain as a string.
+
+            picture_https_url: HTTPS url for the picture image file.
+        """
+        if not self._can_load_media_repo:
+            return
+
+        try:
+            uid = UserID.from_string(user_id)
+
+            def allowed_mime_type(content_type: str) -> bool:
+                if (
+                    self._profile_handler.allowed_avatar_mimetypes
+                    and content_type
+                    not in self._profile_handler.allowed_avatar_mimetypes
+                ):
+                    return False
+                return True
+
+            # download picture, enforcing size limit & mime type check
+            picture = io.BytesIO()
+
+            download_response = await self._http_client.get_file(
+                url=picture_https_url,
+                output_stream=picture,
+                max_size=self._profile_handler.max_avatar_size,
+                is_allowed_content_type=allowed_mime_type,
+            )
+
+            if download_response[3] != 200:
+                raise Exception(
+                    "GET request to download sso avatar image returned {}".format(
+                        download_response[3]
+                    )
+                )
+
+            # upload name includes hash of the image file's content so that we can
+            # easily check if it requires an update or not, the next time user logs in
+            upload_name = "sso_avatar_" + hashlib.md5(picture.read()).hexdigest()
+
+            # bail if user already has the same avatar
+            profile = await self._profile_handler.get_profile(user_id)
+            if profile["avatar_url"] is not None:
+                media_id = profile["avatar_url"].split("/")[-1]
+                media = await self._media_repo.store.get_local_media(media_id)
+                if media is not None and upload_name == media["upload_name"]:
+                    logger.info("skipping saving the user avatar")
+                    return
```

Comment:
If I understand your concern correctly, its possible for the avatar to be a mxc url with media being on a remote server. And just in case, we have a conflicting media ID for that remote file, its better to ensure our code still works. Ok cool!

However, I am having a hard time making sense of that situation in this code section. Because even if there is a conflict, we are only fetching metadata via `get_local_media()` and actual comparison is based on `upload_name` which contains hash of the file itself. So, it would need a match with the upload_name too, making it even more unlikely to happen.

So, what exact change would you like to see here? Just adding server name check at the nested `if` level or something else?

Reply:
I agree it's extremely unlikely. I'm thinking something like
```diff
+ media = None
+ if server_name == self._server_name:
+     media = await self._media_repo.store.get_local_media(media_id)
- media = await self._media_repo.store.get_local_media(media_id)
```
or equivalent

Reply:
Gotcha! Will make that change.

Reply:
Do we need a test for this?

Reply:
I wouldn't bother, since it's so rare.

Reply:
Agree! I will let you take a look at the latest code change and see if its fit to mark as resolved.